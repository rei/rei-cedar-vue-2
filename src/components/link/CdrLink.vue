<template>
  <component
    :is="tag"
    :class="[modifierClass, themeClass]"
    :target="target"
    :rel="computedRel"
    :href="tag === 'a' ? href : null /* don't include the href attribute if not an <a> */"
  >
    <!-- @slot innerHTML on the inside of the anchor component -->
    <slot>Link Text</slot>
  </component>
</template>

<script>
import modifier from 'mixinsdir/modifier';
import themeable from 'mixinsdir/themeable';
/**
 *
 * Cedar 2 component for link.
 *
 * <span class="modifiers">Modifiers</span>
 * {standalone}
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrLink',
  mixins: [modifier, themeable],
  props: {
    tag: {
      type: String,
      default: 'a',
    },
    href: {
      type: String,
      default: '#',
    },
    /** @ignore */
    target: String,
    /** @ignore */
    rel: String,
  },
  computed: {
    baseClass() {
      return 'cdr-link';
    },
    computedRel() {
      if (this.target === '_blank') {
        return this.rel || 'noopener noreferrer';
      }
      return this.rel;
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/vars/CdrLink.vars.pcss';
  @import './styles/CdrLink.pcss';
</style>
