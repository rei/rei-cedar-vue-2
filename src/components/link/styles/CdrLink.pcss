/* ==========================================================================
  # CdrLink

  TOC:

    :Base Styles
      :States
    :Modifiers
      :Style variants
        :Standalone
========================================================================== */

.cdr-link {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  align-items: center;
  background-color: transparent;
  border: 0;
  margin: 0;
  color: $link-color;
  cursor: pointer;
  display: inline-flex;
  fill: $link-color;
  outline: none;
  padding: 0;
  text-decoration: underline;
  vertical-align: top;

  /* States
    ---------- */

  &:hover,
  &:active,
  &:focus {
    color: $link-color;
    outline: none;
    text-decoration: underline;
  }

  &:focus {
    outline: $default-outline;
    outline-offset: 0;
  }
}

/* Style variants
  ========================================================================== */

/* Standalone
    ========== */
.cdr-link--standalone {
  text-decoration: none;

  &:hover,
  &:active,
  &:focus,
  &:visited {
    text-decoration: underline;
  }
}

/* Light Theme */

/* Theme via cascade

:global(.cdr-bg--light) .cdr-link {
  color: $link-color;
  fill: $link-color;

  &:hover,
  &:active,
  &:focus {
    color: $link-color;
  }
}
*/

/* Theme via prop */

:global(.on-light).cdr-link {
  color: $link-color;
  fill: $link-color;

  /* States
    ---------- */

  &:hover,
  &:active,
  &:focus {
    color: $link-color;
  }
}

/* Dark Theme */

/* TODO: get this value */
$link-color-dark: white;

/* Theme via cascade

:global(.cdr-bg--dark) .cdr-link {
  color: $link-color-dark;
  fill: $link-color-dark;
  
  &:hover,
  &:active,
  &:focus {
    color: $link-color-dark;
  }
}
*/

/* Theme via prop */

:global(.on-dark).cdr-link {
  color: $link-color-dark;
  fill: $link-color-dark;

  /* States
    ---------- */

  &:hover,
  &:active,
  &:focus {
    color: $link-color-dark;
  }
}
