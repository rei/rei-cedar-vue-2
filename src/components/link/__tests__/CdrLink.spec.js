import { shallowMount, mount } from '@vue/test-utils';
import sinon from 'sinon';
import CdrLink from 'componentdir/link/CdrLink';
// import CdrThemer from 'componentdir/themer/CdrThemer';

describe('CdrLink', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrLink);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('sets target attr correctly', () => {
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        target: '_self',
      },
    });
    expect(wrapper.attributes().target).toBe('_self');
  });

  it('sets a default href', () => {
    const wrapper = shallowMount(CdrLink);
    expect(wrapper.attributes().href).toBe('#');
  });

  it('sets a default link text', () => {
    const wrapper = shallowMount(CdrLink);
    expect(wrapper.contains('Link Text'));
  });

  it('sets rel attr correctly', () => {
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        rel: 'nofollow',
      },
    });
    expect(wrapper.attributes().rel).toBe('nofollow');
  });

  it('computes target="_blank" rel attr correctly', () => {
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        target: '_blank',
      },
    });
    expect(wrapper.attributes().rel).toBe('noopener noreferrer');
  });

  it('computes the base class correctly', () => {
    const wrapper = shallowMount(CdrLink);
    expect(wrapper.classes()).toContain('cdr-link');
  });

  it('computes classes correctly for standalone modifier', () => {
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        modifier: 'standalone',
      },
    });
    expect(wrapper.classes()).toContain('cdr-link--standalone');
  });

  it('renders a link with a button element and no href attribute', () => {
    const wrapper = mount(CdrLink, {
      propsData: {
        tag: 'button',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('emits a click', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        tag: 'button',
      },
      listeners: {
        'click': spy
      },
    });
    wrapper.trigger('click');
    expect(spy.called).toBeTruthy();
  });

  xit('inherits theme correctly', () => {
    const wrapper = shallowMount(CdrThemer, {
      stubs: {
        'cdr-link': CdrLink,
      },
      slots: {
        default: ['<cdr-link/>']
      },
      propsData: {
        theme: 'dark',
      },
    });

    const link = wrapper.find(CdrLink)

    expect(link.classes()).toContain('on-dark');
  });

  xit('sets theme correctly', () => {
    const wrapper = shallowMount(CdrLink, {
      propsData: {
        theme: 'dark',
      },
    });
    expect(wrapper.classes()).toContain('on-dark');
  });

  it('does nothing if no theme provided', () => {
    const wrapper = shallowMount(CdrLink);
    expect(wrapper.vm.computedTheme).toBeFalsy();
  });
});
