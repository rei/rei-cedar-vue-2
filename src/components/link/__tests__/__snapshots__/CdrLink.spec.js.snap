// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrLink renders a link with a button element and no href attribute 1`] = `
<button
  class="cdr-link"
/>
`;

exports[`CdrLink renders correctly 1`] = `
<a
  class="cdr-link"
  href="#"
/>
`;
